<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Привет еще раз. Спасибо, что доделал работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Привет! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Меня зовут Слепцов Артем и я буду проверять твой проект) Ты проделал большую работу над проектом, однако его еще можно улучшить. Будет здорово, если ты, надеюсь, не против, если я буду на ты, будешь отвечать на комментарии и участвовать в диалоге. 

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках (если фон комментария <font color=green>зелёный</font> - всё сделано правильно. Рекомендации укажу таким же цветом; <font color=orange>оранжевый</font> - некритичные замечания. Если таких замечаний меньше трех - проект может быть принят без их отработки; <font color=red>красный</font> - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<h2> Комментарий ревьюера</h2>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе следует приводить краткий план того, что надо сделать(если этот план имеется), а также информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут. Также название работы лучше отображать в отдельной ячейке и делать крупный шрифт (заголовок). Так работа выглядит презентабельно. Подробнее смотри [методичку](https://www.notion.so/35d9c2f32ffa42f3b17010737aedca4b) по оформлению проектов.

</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd

In [2]:
dt = pd.read_csv('/datasets/data.csv')
dt

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

### Вывод

Из открытой таблицы видно:

**1**. что есть все данные которые позволят ответить на cледующие вопросы:

- есть ли зависимость между наличием детей и возвратом кредита в срок?
- есть ли зависимость между семейным положением и возвратом кредита в срок?
- есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- как разные цели кредита влияют на его возврат в срок?

**2.** заголовки таблицы заполнены корректно;

**3.** интересущие нас стоблбцы : children,  family_status,  income_type,  debt,  purprose.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

У меня имеется несколько комментариев по данному шагу:
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен.

</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
dt.info() # данные по таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
dt.isnull().sum() # количество пропусок


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
dt['days_employed'] = dt['days_employed'].fillna(0) # заменяю пропущенные значения в столбце трудовой стаж на 0


In [6]:
dt[dt['total_income']== 0] # (так как столбик доход является определяющим для исследования, проверяю нет ли
# целенаправленно заполнненных нулевых значений


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [7]:
dt['total_income'] = dt['total_income'].fillna(0) # заменю пропущенные значения на 0


In [8]:
dt.info() # проверяю данные по таблице после проведенных замен


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [9]:
dt.groupby('income_type')['total_income'].max() # проверил максилмальные доходы по категориям, чтоб выбрать тип заполнения пропуска

income_type
безработный        2.027225e+05
в декрете          5.382913e+04
госслужащий        9.104515e+05
компаньон          2.265604e+06
пенсионер          7.351033e+05
предприниматель    4.991631e+05
сотрудник          1.726276e+06
студент            9.820163e+04
Name: total_income, dtype: float64

In [10]:
midian_incom_type = dt.groupby('income_type')['total_income'].median() # Ищем медиану по группам типа занятости
print (midian_incom_type)
def mid_incom(quares):# функция которая возвращает значение медианы, в зависисомости от запроса (тип занятости)
    if quares == 'безработный':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'безработный'), 'total_income'] = midian_incom_type[0] 
    if quares == 'в декрете':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'в декрете'), 'total_income'] = midian_incom_type[1]
    if quares == 'госслужащий':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'госслужащий'), 'total_income'] = midian_incom_type[2]
    if quares == 'компаньон':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'компаньон'), 'total_income'] = midian_incom_type[3]
    if quares == 'предприниматель':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'предприниматель'), 'total_income'] = midian_incom_type[5]
    if quares == 'студент':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'студент'), 'total_income'] = midian_incom_type[7]
    if quares == 'пенсионер':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'пенсионер'), 'total_income'] = midian_incom_type[4]
    if quares == 'сотрудник':
        dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'сотрудник'), 'total_income'] = midian_incom_type[6]




income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        139034.452386
компаньон          162401.351555
пенсионер          110179.690761
предприниматель    249581.572474
сотрудник          133546.457238
студент             98201.625314
Name: total_income, dtype: float64


In [11]:
dt[(dt['income_type']== 'безработный') & (dt['total_income']== 0)] # у категории безработный была очень высокая медиана, проверил ли есть
# нулевые значения в этой категории, замена которых могла сильно исказить значения, их не оказалось

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [12]:
dt['income_type'].apply(mid_incom)

0        None
1        None
2        None
3        None
4        None
         ... 
21520    None
21521    None
21522    None
21523    None
21524    None
Name: income_type, Length: 21525, dtype: object

In [13]:
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'сотрудник'), 'total_income'] = mid_incom('сотрудник') # заменяю пропуски
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'компаньон'), 'total_income'] = mid_incom('компаньон')
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'пенсионер'), 'total_income'] = mid_incom('пенсионер')
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'госслужащий'), 'total_income'] = mid_incom('госслужащий')
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'предприниматель'), 'total_income'] = mid_incom('предприниматель')
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'безработный'), 'total_income'] = mid_incom('безработный')
#dt.loc[(dt['total_income'] == 0) & (dt['income_type'] == 'в декрете'), 'total_income'] = mid_incom('в декрете')
#dt['income_type'].apply(mid_incom)

dt[dt['total_income'] == 0] # проверил замену значений
    
    


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Пустые таблицы на печать лучше не выводить. Стоит вывести размеры получаемой таблицы и показать, что число строк в ней равно 0. В этом тебе поможет аттрибут таблицы shape.

</div>

### Вывод

В столбцах трудовой стаж и ежемесячный доход были обнаружены пропуски, они были количественными переменными.

Пропуски могли возникнуть по следующим причинам:
респонденты не указали данные, 
могли возникнуть технические проблемы,
человеческий фактор (ошибки сотрудников обрабатыващих данные).

Данные по трудовому стажу не требуются для ответа на поставленые задачи и в них есть артефакты в виде отрицательных значений, поэтому пропуски можно заменть на 0 не переживая за результат.
Данные по ежемесячному доходу важный показатель в исследовании. Пропуски  заполнил медианой расчитанной по типам занятости. Решил использовать этот метод, так как расчет по средней завысил бы средний доход в виду наличия больших крайних значений. Расчет же  медианы по группам позволил более точно опредилить доход согласно типа занятости.

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Параметр стажа в анализе мы не используем, поэтому такая его обработка вполне подойдет. Для дохода также все сделано вполне логично. Тип занятости влияет на получаемый доход. Но процесс стоит автоматизировать. Твою функцию mid_incom вполне можно доработать и сделать так, чтобы она сразу заполняла пропуски в данных.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Отлично, код несколько оптимизирован. Однако лучше сделать так, чтобы тип занятости определялся автоматически при заполнении. В этом тебе поможет цикл по уникальным типам занятости. В нем мы можем рассчитывать медианный доход по данному типу занятости и сразу же заполнять пропуски. 

</div>

### Замена типа данных

In [14]:
dt['total_income'] = dt['total_income'].astype(int) # изменяю формать на целочислленный
dt['days_employed'] = dt['days_employed'].astype(int)
dt.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Интересующий столбец для исследования имел вещественный формат flot поэтому для изменения формата на целочисленный intепger подходил метод .astype()

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена только для дохода. Стаж также стоит сделать целочисленным показателем. Также хорошо бы освоить метод to_numeric() для будущих работ. 

</div>

### Обработка дубликатов

In [15]:
dt.duplicated().sum() #считаю полные дубликаты

54

In [16]:
dt = dt.drop_duplicates().reset_index(drop = True) # удаляю полные дубликаты
dt.duplicated().sum()

0

In [17]:
dt['family_status'].value_counts() # проверил  имена

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [18]:
dt['children'].value_counts() # проверил  имена

 0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [19]:
dt['debt'].value_counts()

0    19730
1     1741
Name: debt, dtype: int64

In [20]:
dt['children'].value_counts() # проверил на артефакты столбец дети
dt[dt['children']== -1]['debt'].sum() # посмотрел многоли артефактов попадает на просроченную задолженность
dt.loc[dt['children']== -1, 'children'] = 1 # заменил артефакт, на значение
dt.loc[dt['children']== 20, 'children'] = 2 # заменил артефакт, на значение
dt[dt['children']== -1]['debt'].sum()

0

In [21]:
dt[dt['total_income']<0] # проверил столбец ежемесячный доход на отсутсвие артефактов в виде отрицательных значений

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [22]:
dt['purpose'].value_counts()


свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

### Вывод

В первую очередь я использовал метод для выявления и удаления полных дубликатов. Затем проверил интересующие столбцы с категориальными данными на наличие повторов (количественные столбцы не рассматривал так как наличие дубликатов там допустимо). Из проверенных столбцов дубликаты были выявлены только в столбце цели, но бороться с ними мы буду в разделе лемматизация.
Причины возникновения дубликатов могут быть следующие:
- полные дубликаты могли возникнуть из за технических неисправностей, либо несовершенства программного обеспечения позволяющего подгружать однотипные данные,
- дубликаты же в целях возникли из за возможности свободного ответа на вопрос а не выбора определенной категории

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.
    
Здорово, что ты также заметил и испарвил ошибки в данных.

</div>

### Лемматизация

In [23]:
from pymystem3 import Mystem
m = Mystem()
('').join(m.lemmatize('свадьба, недвижимости, ремонт, жилье, автомобиль, образования')) # выделил целевые слова для категоризации
                    

'свадьба, недвижимость, ремонт, жилье, автомобиль, образование\n'

In [24]:
def changes(purpose):
    words = m.lemmatize(purpose)
    if 'свадьба' in words:        
        return 'свадьба'
    if 'недвижимость' in words:        
        return 'операции с недвижимостью'
    if 'жилье' in words:
        if 'ремонт' in words:
            return 'ремонт жилья'
        return 'операции с недвижимостью'
    if 'автомобиль' in words:        
        return 'приобретение автомобиля' 
    if 'образование' in words:        
        return 'образование'
    
dt['purpose_change'] = dt['purpose'].apply(changes)
dt['purpose_change'].value_counts()       
    
    


операции с недвижимостью    10207
приобретение автомобиля      4308
образование                  4014
свадьба                      2335
ремонт жилья                  607
Name: purpose_change, dtype: int64

### Вывод

Выделил основные группы целей: свадьба, недвижимости, ремонт, автомобиль, образования. Согласно выделенным ключевым словам провел лемматизацию с помощью pymystem3

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Круто, ты сделал самую сложную часть проекта. Использовал нужные методы, и даже провел группировку с помощью собственной функции. Выбор лемм обоснован при помощи подсчета часто встречающихся слов. 

</div>

### Категоризация данных

In [25]:
dt['group_income'] = pd.qcut(dt['total_income'], 8) #разбиваю на группы
dt['group_income'].value_counts() 

(140116.0, 162401.0]      2765
(20666.999, 83890.0]      2685
(254213.75, 2265604.0]    2684
(195767.5, 254213.75]     2684
(127580.25, 140116.0]     2684
(107654.5, 127580.25]     2684
(83890.0, 107654.5]       2683
(162401.0, 195767.5]      2602
Name: group_income, dtype: int64

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Здорово, что ты освоил данный метод категоризации. Так группы получаются одинакового размера, и мы можем по ним построить выводы. 

</div>

In [26]:
dict_family_status = dt[['family_status_id','family_status']].drop_duplicates().reset_index(drop = True) # словарь семейный статус

dt['education'] = dt['education'].str.lower()
dict_education = dt[['education_id','education']]
dict_education =dict_education.drop_duplicates().reset_index(drop = True) # словарь образование
print (dict_family_status)
dict_education

   family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем


education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Словари выделены. Теперь мы можем убрать текстовые параметры из основной таблицы, что увеличит эффективность работы с данными. 

</div>

### Вывод

Столбец total_incom был заполнен количественными данными, чтоб провести исследования нужно было провести его группирвоку. Группы  сдеал методом qcut, который разбивает данные на равные части. Разделить решил на 8 частей. Такой разделитель мне кажется оптимальный, так группы получаются не  мелкие и данных хватает для вывода, и не крупные чтоб затереть закономерности.

Выдели два словаря: словарь образование, словарь семейное положение. У каждой категории в этих словарях есть свой уникальный id, по которуму можно работать с данными в будущем. 


<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Исходя из чего выбран именно такой способ категоризации столбца с доходами? Не во всех группах содержится достаточное число клиентов для построения по ним надежных выводов. Лучше использовать разбиение по квантилям данных (метод qcut). Тогда группы получатся одинакового размера. Или же можно узнать в интернете процентный состав общества по доходам и затем уже категоризовать столбец с доходами по найденному соотношению с помощью персентилей. 

</div>

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

В данных есть словари. Попробуй их обнаружить и выделить. Это довольно распространенный способ хранения информации. Мне важно понять, что он тобой освоен. Заметь, что в основной таблице мы имеем по 2 параметра с образованием и семейным статусом. Именно по ним и стоит выделить словари. 

</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [27]:
total_debt = dt['debt'].sum()
dt_children = dt.groupby(['children']).agg({'debt': ['sum','count']})
dt_children['conv'] = (dt_children['debt']['sum'] /dt_children['debt']['count']   * 100).round(2)
dt_children.sort_values('conv')


debt         conv
           sum  count      
children                   
5            0      9  0.00
0         1063  14107  7.54
3           27    330  8.18
1          445   4856  9.16
2          202   2128  9.49
4            4     41  9.76

### Вывод

Исследование данных показало, что есть зависимость между наличием детей и возратом кредита в срок, самые пунктуальные кредиторы оказась без детей их доля просрочки к выданным кредитам составила 7,54 %. Конечно есть семьи с 5 детьми у которых нет вообще просроки, но вывод по ним делать сложно, так выборка очень маленькая всего  9 респондентов.
Самая же большая доля по просрочке у людей с 4 детьми их доля составила 9,76 %.

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

20 детей - это все же ошибка в данных, которую надо было исправить на этапе предобработки данных. Также есть и другие ошибки. Обнаружить их тебе помогут методы describe и value_counts. 

</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [28]:
dt_family = dt.pivot_table(index = ['family_status'], columns =  'debt', values = 'family_status_id', aggfunc = 'count') #провел группировку pivot_table 
print (dt_family)  
dt_family= dt.groupby(['family_status']).agg({'debt': ['sum','count']})
dt_family['conv'] = (dt_family['debt']['sum'] /dt_family['debt']['count']   * 100).round(2)
dt_family.sort_values('conv')

debt                       0    1
family_status                    
Не женат / не замужем   2536  274
в разводе               1110   85
вдовец / вдова           896   63
гражданский брак        3775  388
женат / замужем        11413  931


debt         conv
                       sum  count      
family_status                          
вдовец / вдова          63    959  6.57
в разводе               85   1195  7.11
женат / замужем        931  12344  7.54
гражданский брак       388   4163  9.32
Не женат / не замужем  274   2810  9.75

### Вывод

Исследование данных показало, что есть зависимость между семейным положением и возратом кредита в срок, самыми пунктуальными кредиторами оказась люди с непростыми жизненными ситуациями: кто столкнулся со смертью близких и у кого распался брак, их доли соответственно  6,57 % и 7,11 %. Люди которые же не когда не были скованы узами брака чаще всего допускают просрочку 9,75 %. 


Провел  группировки с помощью pivot_table, но в данном задании мне кажется его использование не удобно. И дальнейшие расчеты сделал другигим методом. pivot_table такой метод таблицы называется «широкий».Он удобен, когда нужно сравнить значения столбцов. Методы groupby() и agg() удобны, когда нужно применить функцию к столбцу со сгруппированными  данными или создать новый столбец на их основе.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Вывод не противоречит полученному результату. 

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
dt_income= dt.groupby(['group_income']).agg({'debt': ['sum','count']})
dt_income['conv'] = (dt_income['debt']['sum'] /dt_income['debt']['count']   * 100).round(2)
dt_income.sort_values('conv')

debt        conv
                        sum count      
group_income                           
(254213.75, 2265604.0]  185  2684  6.89
(195767.5, 254213.75]   198  2684  7.38
(20666.999, 83890.0]    206  2685  7.67
(83890.0, 107654.5]     221  2683  8.24
(140116.0, 162401.0]    231  2765  8.35
(127580.25, 140116.0]   226  2684  8.42
(107654.5, 127580.25]   235  2684  8.76
(162401.0, 195767.5]    239  2602  9.19

### Вывод

Проведенное исследование не выявило явной линейной зависимости между доходом и возратом кредита срок. Как самыми наденжными так самыми ненадежными являются люди с высокими доходами. Но все же рейтинг надежных заемщиков возглавляют 2 категрии с самыми высокими доходами(доходы свыше 195 тыс). Что касается 3 категории дохода, которая оказалась самой ненадежной, думаю это связано с тем, что они люди с очень высокий зарплатой, но не достаточно богатыми чтоб позволить все радости мира, и не расчитав свои силы чаще чем другие залазеют в долги, чтоб удовлетворить свои растущие потребности, которые не покрывают их доходы.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Посмотри на результаты при ином способе категоризации. 

</div>

- Как разные цели кредита влияют на его возврат в срок?

In [30]:
dt_purpose= dt.groupby(['purpose_change']).agg({'debt': ['sum','count']})
dt_purpose['conv'] = (dt_purpose['debt']['sum'] /dt_purpose['debt']['count']   * 100).round(2)
dt_purpose.sort_values(['conv'])



debt         conv
                          sum  count      
purpose_change                            
ремонт жилья               35    607  5.77
операции с недвижимостью  747  10207  7.32
свадьба                   186   2335  7.97
образование               370   4014  9.22
приобретение автомобиля   403   4308  9.35

### Вывод

Исследование опказало что про люди берущие кредит на автомобиль и образование чаще просрачивают кредит( 9,35 % и 9,22 %), самыми же надежнымм кредиторами являются клиенты цель кредита которых: ремонт жилья и операции с недвижимостью ( 5,77 % и 7,46%).

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

На данном шаге рекомендуется использовать метод сводных таблиц (pivot_table()). Однако он является аналогом метода groupby(). Изучи метод сводных таблиц и примени его хотя бы в одном пункте анализа.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Метод сводных таблиц использован. Данный метод нам сильно пригодится в дальнейших проектах. Только в таблице ты не получил доли должников. Для этого надо столбей 1 поделить на сумму столбцов 0 и 1, т.е. на общее число клиентов данной группы.  

</div>

### Шаг 4. Общий вывод

В ходе изучени исследовния  были проведены следущие работы:

**Этап 1** Загрузка данных:
- загрузил данные с помощью метода read_csv();
- проверил есть ли данные позволящие провести исследование;
- проверил заголовки таблицы
- получил данные по таблице с помощью метода info().

**Этап 2** Предобработка данных:
- обнаружил в интересующих данных пропуски, все они имели формат Nan, поэтом все заменил на 0с помощью метод fillna();
- для столбца доход написал функцию которая заменяет, нули на медианы расчитанные по источнику дохода;
- заменил артефакты в столбце дети;
- использовал pymystem3 для лемматизация, и выдел категирии в целях получения кредита;
- провел категоризацию столбца доход, для это применил метод qcut();
- выдели два словаря: по группу образование и семейное положение.

**Этап 3** Выводы
- Провел  анализ показателей  с помощью pivot_table groupby() и agg()

В ходе работы надо было ответить на следущие вопросы:

- есть ли зависимость между наличием детей и возвратом кредита в срок?
- есть ли зависимость между семейным положением и возвратом кредита в срок?
- есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- как разные цели кредита влияют на его возврат в срок?

**Зависимость между наличием детей и возвратом кредита в срок.**

На основе данных были определенны 5 основных категорий по количеству детей: 0,1,2,3,4.

**Результаты исследования:**
- 0 детей: количество выданных кредитов 14107, количество просроченных кредитов 1063, доля просрочки по кредитам 7,54 %
- 1 ребенок: количество выданных кредитов 4856, количество просроченных кредитов 445, доля просрочки по кредитам 9,16 %
- 2 ребенка: количество выданных кредитов 2128, количество просроченных кредитов 202, доля просрочки по кредитам 9,49 %
- 3 ребенка: количество  выданных кредитов 330, количество просроченных кредитов 27, доля просрочки по кредитам 8,18 %
- 4 ребенка: количество  выданных кредитов 41, количество просроченных кредитов 4, доля просрочки по кредитам 9,76 %

Исследование данных показало, что есть зависимость между наличием детей и возратом кредита в срок, самые пунктуальные кредиторы оказась без детей их доля просрочки к выданным кредитам составила 7,54 %. Самая же большая доля по просрочке у людей с 4 детьми их доля составила 9,76 %. Это в принципе логично каждый ребенок это значетильная нагрузка на бюджет семьи, и чем ниже доход у семей с детьми тем тяжелее им приходится с выплатой кредитов и каждай непредвиденная трата может выбить их из коллеии.

**Зависимость между семейным положением и возвратом кредита в срок.**

На основе данных были определенны 5 основных категорий по количеству детей:

- вдовец / вдова	
- в разводе	
- женат / замужем
- гражданский брак
- Не женат / не замужем

**Результаты исследования:**
- вдовец / вдова: количество выданных кредитов 959, количество просроченных кредитов 63, доля просрочки по кредитам 6,57 %
- в разводе: количество выданных кредитов 1195, количество просроченных кредитов 85, доля просрочки по кредитам 7,11 %
- женат / замужем: количество выданных кредитов 12344, количество просроченных кредитов 931, доля просрочки по кредитам 7,54 %
- гражданский брак: количество  выданных кредитов 4163, количество просроченных кредитов 388, доля просрочки по кредитам 9,32 %
- Не женат / не замужем: количество  выданных кредитов 2810, количество просроченных кредитов 274, доля просрочки по кредитам 9,75 %

Исследование данных показало, что есть зависимость между семейным положением и возратом кредита в срок, самыми пунктуальными кредиторами оказась люди с непростыми жизненными ситуациями: кто столкнулся со смертью близких и у кого распался брак, их доли соответственно  6,57 % и 7,11 %. Люди которые же не когда не были скованы узами брака чаще всего допускают просрочку 9,75 %, 
это видимо происходит из за того что у  них доля личной ответсвенности ниже, живут более своими потребностями.

**Зависимость между уровнем дохода и возвратом кредита в срок.**

На основе данных были выделены 8  категорий дохода:
- от 20700 до 83900: 
- от 83900 до 107650: 
- от 107650 до 127580: 
- от 127580 до 140100:
- от 140100 до 162400: 
- от 162400 до 195770:
- от 195770 до 254210: 
- от 254210 до 2265604: 

**Результаты исследования:**
- от 20700 до 83900: количество выданных кредитов 2685, количество просроченных кредитов 206, доля просрочки по кредитам 7,67 %
- от 83900 до 107650: количество выданных кредитов 2683, количество просроченных кредитов 221, доля просрочки по кредитам 8,24 %
- от 107650 до 127580: количество выданных кредитов 2684, количество просроченных кредитов 235, доля просрочки по кредитам 8,76 %
- от 127580 до 140100: количество  выданных кредитов 2684, количество просроченных кредитов 226, доля просрочки по кредитам 8,42 %
- от 140100 до 162400: количество  выданных кредитов 2765, количество просроченных кредитов 231, доля просрочки по кредитам 8,35 %
- от 162400 до 195770: количество  выданных кредитов 2602, количество просроченных кредитов 239, доля просрочки по кредитам 9,19 %
- от 195770 до 254210: количество  выданных кредитов 2684, количество просроченных кредитов 198, доля просрочки по кредитам 7,38 %
- от 254210 до 2265604: количество  выданных кредитов 2684, количество просроченных кредитов 185, доля просрочки по кредитам 6,89 %


Проведенное исследование не выявило явной линейной зависимости между доходом и возратом кредита срок. Как самыми наденжными так самыми ненадежными являются люди с высокими доходами. Но все же рейтинг надежных заемщиков возглавляют 2 категрии с самыми высокими доходами(доходы свыше 195 тыс). Что касается 3 категории дохода, которая оказалась самой ненадежной, думаю это связано с тем, что они люди с очень высокий зарплатой, но не достаточно богатыми чтоб позволить все радости мира, и не расчитав свои силы чаще чем другие залазеют в долги, чтоб удовлетворить свои растущие потребности, которые не покрывают их доходы.

**Зависимость между целями кредита и возвратом кредита в срок.**

На основе данных были определенны 5 основных категорий целей кредитования: 

- ремонт жилья;
- операции с недвижимостью;
- свадьба;
- образование;
- приобретение автомобиля.

**Результаты исследования:**
- ремонт жилья: количество выданных кредитов 607, количество просроченных кредитов 35, доля просрочки по кредитам 5,77 %
- операции с недвижимостью: количество выданных кредитов 10207, количество просроченных кредитов 747, доля проскрочки по кредитам 7,32 %
- свадьба: количество выданных кредитов 2335, количество просроченных кредитов 186, доля просрочки по кредитам 7,97 %
- образование: количество  выданных кредитов 4014, количество просроченных кредитов 370, доля просрочки по кредитам 9,22 %
- приобретение автомобиля: количество  выданных кредитов 4308, количество просроченных кредитов 403, доля просрочки по кредитам 9,35 %

Исследование показало, что люди берущие кредит на автомобиль и образование чаще просрачивают кредит( 9,35 % и 9,22 %), самыми же надежнымм кредиторами являются клиенты цель кредита которых: ремонт жилья и операции с недвижимостью ( 5,77 % и 7,46%). В текущих данных есть определенная закономерность, клиенты которые берут кредит на недвижимость здраво оценивает свое финнасовое состояние и возможности, да и недвижимость зачастую является обремененной кредитором, и чтоб не было ни каких проблем заемщики стараюся гасить в срок. Что же касается кредита на авто, то данные тоже объсними, люди зачастую оценивают машину как игрушку, и многие пытаются приобрести её, даже если не позволяют финансы.




Cогласно полученным данным можно сформировать следущие факторы :
факторы надежности:
1 наличие детей : нет
2 семейный статус: вдовец,вдова; в разводе
3 уровень дохода: свыше 195
4 цель: ремонт, недвижимость

факторы риска:
1 наличие детей: 2 и более
2 не замужем/ не женат; в гражданском браке
3 доход от 162 до 195
4 цель: автомобиль, образование

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. В нем можно приводить полученные в ходе работы значения. Также можно расписать все, что было сделано в работе.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, правильно. Здорово, что по каждому пункту работы приведен вывод. 

</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

- В начале работы следует описывать данные, с которыми работаешь;
- Категоризацию по доходу стоит делать иначе. Смотри данный пункт подробнее;
- Выдели словари в данных;
- Изучи и используй в анализе метод сводных таблиц;
- Финальный вывод стоит писать подробно.

Ты проделал большую работу, молодец! Исправь отмеченные выше помарки. Также есть достаточное число желтых комментариев. Думаю, исправить их тебе не составит труда. Жду твою работу :)

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Помарки исправлены, и теперь работа выполнена хорошо. Ты здорово поработал над исправлениями, молодец. Успехов в дальнейших проектах :)

</div>
   

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.